File: epi_judge_java/epi/TwoSortedArraysMerge.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;

public class TwoSortedArraysMerge {

    public static void mergeTwoSortedArrays(List<Integer> a, int m, List<Integer> b, int n) {
        int writeIdx = m + n - 1;
        m--;
        n--;
        while (m >= 0 || n >= 0) {
            if (n < 0) break;
            if (m < 0 || a.get(m) < b.get(n)) {
                a.set(writeIdx, b.get(n));
                n--;
            } else {
                a.set(writeIdx, a.get(m));
                m--;
            }
            writeIdx--;
        }
    }

    @EpiTest(testDataFile = "two_sorted_arrays_merge.tsv")
    public static List<Integer>
    mergeTwoSortedArraysWrapper(List<Integer> A, int m, List<Integer> B, int n) {
        mergeTwoSortedArrays(A, m, B, n);
        return A;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "TwoSortedArraysMerge.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/two_sorted_arrays_merge.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

void MergeTwoSortedArrays(vector<int> &nums1, int m, const vector<int> &nums2, int n) {
    int i1 = m - 1;
    int i2 = n - 1;
    while (i1 >= 0 || i2 >= 0) {
        if (i1 == -1) {
            nums1[i2] = nums2[i2];
            i2--;
        } else if (i2 == -1) {
            break;
        } else if (nums1[i1] >= nums2[i2]) {
            nums1[i1 + i2 + 1] = nums1[i1];
            i1--;
        } else {
            nums1[i1 + i2 + 1] = nums2[i2];
            i2--;
        }
    }
}

vector<int> MergeTwoSortedArraysWrapper(vector<int> A, int m,
                                        const vector<int> &B, int n) {
    MergeTwoSortedArrays(A, m, B, n);
    return A;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"A", "m", "B", "n"};
    return GenericTestMain(
        args, "two_sorted_arrays_merge.cc", "two_sorted_arrays_merge.tsv",
        &MergeTwoSortedArraysWrapper, DefaultComparator{}, param_names);
}
